sources/web/datalab/polymer/modules/ApiManager.ts

```diff
@@ -421,17 +437,34 @@ class ApiManager {
     const successCode = options.successCode || 200;
     const request = new XMLHttpRequest();
     const noCache = options.noCache || false;
+    const failureCodes = options.failureCodes;
 
     return new Promise((resolve, reject) => {
       request.onreadystatechange = () => {
         if (request.readyState === 4) {
           if (request.status === successCode) {
+
+            // If this is the first success after failures, call the connected handler
```

Comment:
This is what we have here. This is the only place where XHRs are explicitly used, then it's encapsulated in `ApiManager._xhrTextAsync`.